src/bgg/error_simulator.rs

```diff
@@ -0,0 +1,384 @@
+use super::{circuit::PolyCircuit, Evaluable};
+use crate::poly::{Poly, PolyElem};
+use itertools::Itertools;
+use num_bigint::BigUint;
+use num_traits::{One, Zero};
+use std::ops::{Add, Mul, Sub};
+
+impl<P: Poly> PolyCircuit<P> {
+    pub fn simulate_error(&self, dim: u32) -> Vec<MPolyCoeffs> {
+        let one = ErrorSimulator::default_from_dim(dim);
+        let inputs = vec![ErrorSimulator::default_from_dim(dim); self.num_input()];
+        let outputs = self.eval(&(), one, &inputs);
+        outputs.into_iter().map(|output| output.h_norm).collect_vec()
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct ErrorSimulator {
+    pub h_norm: MPolyCoeffs,
+    pub plaintext_norm: BigUint,
+    pub dim: u32,
+}
+
+impl ErrorSimulator {
+    pub fn new(h_norm: MPolyCoeffs, plaintext_norm: BigUint, dim: u32) -> Self {
+        Self { h_norm, plaintext_norm, dim }
+    }
+
+    pub fn default_from_dim(dim: u32) -> Self {
+        let h_norm = MPolyCoeffs::new(vec![BigUint::from(dim)]);
+        let plaintext_norm = BigUint::one();
+        Self { h_norm, plaintext_norm, dim }
+    }
+}
+
+impl Add<Self> for ErrorSimulator {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self::Output {
+        self + &rhs
+    }
+}
+
+impl<'a> Add<&'a Self> for ErrorSimulator {
```

Comment:
```suggestion
impl Add<&Self> for ErrorSimulator {
```
general comment for all `'a` it's needless (<https://rust-lang.github.io/rust-clippy/master/index.html#needless_lifetimes>)